nupic/research/frameworks/pytorch/hooks/hook_manager.py

```diff
@@ -0,0 +1,169 @@
+
+        # Register the hooks via class method.
+        tracked_vals = self.register_storage_hooks(model, hook_class,
+                                                   hook_type=hook_type,
+                                                   include_modules=include_modules,
+                                                   include_names=include_names,
+                                                   include_patterns=include_patterns)
+
+        # These are the functions that called every forward or backward pass.
+        self.hooks = tracked_vals[0]
+
+        # These are handles to the hooks; PyTorch lets the user unregister
+        # hooks through these handles.
+        self._hook_handles = tracked_vals[1]
+
+        # These are the filtered modules that will be tracked.
+        self.tracked_modules = tracked_vals[2]
+
+    def __enter__(self):
+        """Start tracking when `with` is called."""
+        self.start_tracking()
+        return self
+
+    def __exit__(self, *args):
+        """Stop tracking when `with` block is left."""
+        self.stop_tracking()
+
+    @classmethod
+    def register_storage_hooks(
+        cls, model, hook_class,
+        hook_type="forward",
+        include_modules=None,
+        include_names=None,
+        include_patterns=None,
+    ):
+        """
+        Register hooks on the model's filtered modules. The default is to track all
+        modules.
+
+        :param model: model to add hooks to
+        :param hook_class: class subclassed from `TrackStatsHookBase`
+        :param hook_type: whether to register the hook as "forward" or "backward"
+        :param include_modules: a list of module types to track
+        :param include_names: a list of module names to track e.g. "features.stem"
+        :param include_patterns: a list of regex patterns to include by name
+        """
+        assert hook_type in ["forward", "backward"]
+
+        hooks = []
+        handles = []
+        tracked_modules = dict()
+
+        # Filter out the modules so that only some of them are tracked.
+        named_modules = filter_modules(model,
+                                       include_modules=include_modules,
+                                       include_names=include_names,
+                                       include_patterns=include_patterns)
+
+        # By default, track all modules.
+        named_modules = named_modules or named_modules
+
+        # Register hooks on the modules.
+        for n, m in named_modules.items():
+
+            hook = hook_class(name=n)
+            if hook_type == "forward":
+                handle = m.register_forward_hook(hook)
+            else:
+                handle = m.register_backward_hook(hook)
+
+            hooks.append(hook)
+            handles.append(handle)
+            tracked_modules[n] = m
+
+        return hooks, handles, tracked_modules
+
+    def start_tracking(self):
+        for hook in self.hooks:
```

Comment:
If I want to start and stop some hooks at different times, do I break it down into several HookManagers?

Reply:
That's the idea. The managers should be easy enough to work with that you can create as many as needed, ideally.